mobile/library/cc/stream_callbacks.cc

```diff
@@ -10,9 +10,16 @@ namespace Platform {
 
 namespace {
 
+// StreamCallbacksWrapper allows us to create a raw pointer instance that wraps the StreamCallbacks
+// shared_ptr, so we can store it in the context as a void*.
+struct StreamCallbacksWrapper {
+  StreamCallbacksSharedPtr stream_callbacks;
+};
+
 void c_on_headers(envoy_headers headers, bool end_stream, envoy_stream_intel intel, void* context) {
-  auto stream_callbacks = *static_cast<StreamCallbacksSharedPtr*>(context);
-  if (stream_callbacks->on_headers.has_value()) {
+  StreamCallbacks& stream_callbacks =
+      *static_cast<StreamCallbacksWrapper*>(context)->stream_callbacks;
```

Comment:
nit: Totally your call, but I had a slightly hard time interpreting what this line does.  There are a few similar lines. Would it make sense to add a helper function which is basically:

```
StreamCallbacks& getStreamCallbacksFromContext(void* context) {
   return static_cast<StreamCallbacksWrapper*>(context)->stream_callbacks;
}
```
which could have a nice hefty comment about how we store a `void*` pointer to a `StreamCallbacksWrapper`, which is really just a way of saying we're storing a heap-allocated `StreamCallbacksSharedPtr`?